### DISCUSSION TITLE
Network adaptive media rate control

### DISCUSSION THREAD

**Discussion by WangHoi** [rustdesk/rustdesk#3981] [Ideas]
Take a glimpse at code, it seems that screen capture and encoded without network adaptive rate control.
Maybe we can improve UX with constrained quality vp9 encoding and bandwidth estimation(congestion control).

**Comment by WangHoi** [comment]
First need to investigate network simulation tools.
Then we can try something, eg. SCReAM algorithm (RFC 8298)

Congestion Control for RTP Media: a Comparison on Simulated Environment
https://arxiv.org/pdf/1809.00304.pdf

**Comment by rustdesk** [comment]
@WangHoi Could you contribute?

**Comment by rustdesk** [comment]
@fufesou added a simple fix https://github.com/rustdesk/rustdesk/pull/322 for #295 under weak network, but it is far away from your congestion control.

**Comment by simpzan** [comment]
Why not use webrtc for media transportation? Congestion control is built-in.

**Comment by simpzan** [comment]
What is the current solution for media transportation? just curious.

**Comment by fufesou** [comment]
> Why not use webrtc for media transportation? Congestion control is built-in.

Yes, webrtc may be a good choice. But we are not familiar with it for now. Maybe in the future, we'll use it.

For now, media transportation is implemented by transferring video and audio frames.

Audio frames are very small, and no need to be considered.
Video frames are much bigger. Only one key frame(much bigger than other video frames) is transffered when connection is established.

Feedback is used to avoid congestion. That is, video frames are captured and stored in a buffer queue.
A new video frame can be captured and stored into the queue, only when video frame is fetched by the called.

**Comment by fufesou** [comment]
> What is the current solution for media transportation? just curious.

![image](https://user-images.githubusercontent.com/13586388/148713150-02cd7a07-d380-4956-80c9-f1006d128dee.png)

**Comment by rustdesk** [comment]
We may need to dive into vpx rate control library 

https://github.com/webmproject/libvpx/blob/16837ae1680bbc73381570cc783439b0ea121ba6/CHANGELOG#L10
https://github.com/webmproject/libvpx/blob/16837ae1680bbc73381570cc783439b0ea121ba6/CHANGELOG#L70
https://github.com/webmproject/libvpx/blob/16837ae1680bbc73381570cc783439b0ea121ba6/CHANGELOG#L79

**Comment by rustdesk** [comment]
https://spec.oneapi.io/versions/latest/elements/oneVPL/source/appendix/VPL_apnds_c.html

**Comment by WangHoi** [comment]
@rustdesk 
I've investigate this interest problem, and like to share some progress:

1. We can measure the capture, encode time, estimate send time, and start capture next frame adaptively, better than fixed 30/10/5 fps.
2. If detected large updated region, or found encoded frame is too large, need to use large QP immediately(re-encode if dectect large frame after encoding), to avoid large frame, and topoff the picture quality lately.
3. Use screen content coding tech, intel mfx support HEVC-SCC, or China's AVS3-SCC standard, or AV1-SCC(Aurora1 Properitary codec).

**Comment by rustdesk** [comment]
@WangHoi Thanks

**Comment by rustdesk** [comment]
@WangHoi 
1. We have done. https://github.com/rustdesk/rustdesk/blob/master/src/server/video_service.rs#L304

**Comment by Speedkore** [comment]
Hi guys, I don't know if this is related but before opening a new issue, I want to expose my problem with low upload bandwidth (less that 1 Mbit). The issue is that when im sharing my screen with another partner, if I move big portions of the screen like scrolling a page of code , then my partner sees all the movements with a big delay (like >10s), until things stabilise.

I'm guessing that there is buffer and rustdesk wants to send it all, but if think it is more important the latency and it will be better to drop frames that are too old that trying to keep a frame rate or something like this.

**Comment by rustdesk** [comment]
Of course, we know this.

**Comment by mon-jai** [comment]
It seems that rustdesk only loads another frame after it completed loading the previous frame. In slow connections, that makes rustdesk really unresponsive.

If it takes 1 second to load a frame, that means that rustdesk will load the first frame at 0:01, the second frame at 0:02 and the 30th frame at 0:30, **which is the last frame of the first second and 29 seconds outdated**.

Can we improve rustdesk so that it could drop outdated frame and always load the latest frame available?
